Add design doc for Application Identity validation

orenbm opened the pull request:
When we add a new authenticator to Conjur, we will need to add the mechanism
of validating the host's Application Identity. Currently we have no generic
way to do that and because of that it is less straight-forward to add a new
authenticator.

This doc tackles this issue and suggests a new approach.

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+
+##### Note
+
+In this doc I talk about authenticating hosts with Conjur. Everything I say
```
In this doc, I will cover authenticating hosts with Conjur. Everything about authenticating hosts is also relevant to users. For the scope of this document, `host`, `user` or `role` are interchangeable"

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+
+In this doc I talk about authenticating hosts with Conjur. Everything I say
+ about authenticating hosts is relevant also for authenticating users
+ . Whenever I say one of `host`, `user` or `role` it has the same meaning.
```
This preamble is very useful. Like that you added it

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+  
+## Introduction
+
+In several authenticators we require hosts to contain some data which define
```
in several or in all? I would list them in (). Example: (authn-k8s, authn-azure, ...)

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+   - System-Assigned Identity (optional)
+   - User-Assigned Identity (optional)
+ 
+A host that will authenticate using the Azure authenticator will be defined
```
will be defined in its annotations as follows:

sigalsax commented on design/authenticators/application-identity/application-identity.md:
````diff
@@ -0,0 +1,464 @@
+    authn-azure/resource-group: test-group 
+``` 
+
+In the example above, the Application Identity is stored as annotations.
````
remove this sentence if you implement the previous comment

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+generic mechanism, as can be seen in the implementations for the [Azure
+Authenticator](/app/domain/authentication/authn_azure/validate_application_identity.rb) and 
+[Kubernetes Authenticator](/app/domain/authentication/authn_k8s/validate_application_identity.rb), where a lot of code is duplicated with
+ minor changes. 
```
why is this on a new line?

orenbm replied:
auto-indent. i disabled it.

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+ Identities are related across authenticators, and how they are different.
+ 
+At this point, to ease the problem, I will act as if Application Identities
+can be defined only in the host's annotations, and not in the ID. In future
```
For the sake of simplicity, from here and throughout the doc, Application Identities will be defined only in the host's annotations, and not in the ID.

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+and then builds a `constraints` hash from the annotations. 
+
+This is the case also in [`authn-k8s`](/app/domain/authentication/authn_k8s/application_identity.rb), with only one difference as the K8s Authenticator
+also looks in the host's ID. 
```
why new line?

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+
+Putting that aside we can see that the behavior is exactly the same. 
+In this case the general class will be defined as follows:
+- input
```
align this line with rest of points

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+   
+#### Validate the Application Identity is defined correctly
+
+In the Azure Authenticator and in the K8s Authenticator we have some
```
remove "some". fluff

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+- output
+  - Nothing if validation succeeded, InvalidApplicationIdentity error if it
+   failed
+   
```
the previous sections are all very very comprehensive. Great job explaining!

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+  - Authenticator name (e.g 'authn-azure')
+  - Authenticator's service-id (e.g 'prod')
+  - host's annotations
+  - Different lists of annotation names
```
A list of annotation names that can be a combination of any of the following:

orenbm replied:
the input won't have one list that is a combination of the following. It will have 3 lists of annotations.

orenbm replied:
They can also be empty if the authenticator doesn't have such a constraint. Because we iterate over the list then if the list is empty -> no error will be raised.

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+    - permitted
+    - required
+    - non-permitted combinations
+  - Authentication request Application Identity object
```
this is the request uri right?

orenbm replied:
what do you mean? I am referring the the Application Identity that is extracted from the Authentication request. For example, in `authn-azure` it is the identity that is extracted from the azure access token

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+     - Authentication request Application Identity object given in the input
+      to this class
+      
+Let's see how the classes will look, in high level.
```
"at a high level"

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+  - Authentication request Application Identity object
+  
+It will perform the following steps:
+  1. Build the Host Application Identity object using the:
```
Here do you mean a completely, separate object that will have the attributes listed below?
In other words an object that has an authenticator name, service-id, host-annotations attributes will be dynamically filled based on the incoming request and the list of permitted annotations predefined?
So before runtime, the Azure Authentication Identity object would look like:
```
    authenticator name: null
    authenticator service-id: null
    host's annotations: null
    list of permitted annotation: subscription id, resource group, etc.
```
and at runtime:
```
    authenticator name: authn-azure
    authenticator service-id: oren-azure
    host's annotations: subscription-id: my-great-subscription-id, resource-group: my-resource-group
    list of permitted annotation: subscription id, resource group, etc.
```

orenbm replied:
No. this class is described above like this:
```
- input
  - authenticator name
  - authenticator service-id
  - annotations
  - List of relevant annotations (e.g `[subscription-id, resource-group
  , system-assigned-identity, user-assigned-identity]
  ` for `authn
  -azure`)
- output
  - `constraints` hash.
```

It will have only one attribute - `constraints`

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+
+The input for this class will be built in each authenticator as it differs
+ between them. For simplicity, the class will receive the `constraints` hash
+  as an input. We should challenge that to find a better way.
```
what other options do we have? a separate object?

orenbm replied:
yes. i'll add that.

orenbm replied:
oh no, i thought you meant something else. I actually don't have yet a better idea but it seems that we can find one if we look for it.

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+
+      def token_identity_from_claims
+        @token_identity = {
+          subscription_id: xms_mirid.subscriptions,
```
where does `xms_mirid` come from if not as input?

orenbm replied:
it does but it's not very interesting. I'll add a comment to clarify

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+                                         system-assigned-identity
+                                       )
+          authentication_request_application_identity: @authentication_request_application_identity_class.new(
+            constraints: @token_identity
```
should this be
`constraints: token_identity_from_claims`? b/c on first run, token_identity will be null and nothing calls `token_identity_from_claims` so why not trigger the loading of the `token_identity` by replacing with what i suggested unless you intend for it to be different?

sigalsax replied:
or will that be filled as part of the `<perform other authentication steps>` calls?

orenbm replied:
you are correct, i missed it. Changed it now.

sigalsax commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+    verify_application_identity_matches_authentication_request
+  end
+
+  def extract_application_identity_from_host
```
Correct me if I am wrong but I feel like `extract_application_identity_from_host` is not really representative of what this function is doing. `.new` just initializes the object but doesn't extract anything. Only host_application_identity.constraints "extracts" and checks if the annotations present are permitted

Maybe `initialize_application_identity_from_host`?

orenbm replied:
i don't want to go into this implementation detail.

orenbm replied:
but you are correct :)

orenbm commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,466 @@
+  ` for `authn
+  -azure`)
+- output
+  - `constraints` hash.
```
@jonahx is there a better option than a hash? Please keep this question in mind while reading the rest of the doc and answer it once you're finished.

orenbm commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,466 @@
+interface is the same, they are not entirely identical. 
+
+At first, the `constraints` method will return a simple key-value hash. We
+should challenge that and find a better way to do it in Ruby.
```
@jonahx any idea?

jvanderhoof replied:
The dangerous way to do this is with `method_missing` (powerful, but treat with great care):
```
➜  dap-wiki git:(master) irb
2.5.1 :001 > class Foo
2.5.1 :002?>     def initialize()
2.5.1 :003?>         @attributes = {
2.5.1 :004 >               bing: 'bang',
2.5.1 :005 >               ping: 'pong'
2.5.1 :006?>           }
2.5.1 :007?>       end
2.5.1 :008?>
2.5.1 :009?>     def method_missing(m, *args, &blk)
2.5.1 :010?>         if @attributes.key?(m.to_sym)
2.5.1 :011?>             @attributes[m.to_sym]
2.5.1 :012?>           else
2.5.1 :013?>             super
2.5.1 :014?>           end
2.5.1 :015?>       end
2.5.1 :016?>   end
 => :method_missing
2.5.1 :017 > foo = Foo.new
 => #<Foo:0x00007f9e74186f40 @attributes={:bing=>"bang", :ping=>"pong"}>
2.5.1 :018 > foo.ping
 => "pong"
2.5.1 :019 > foo.baz
Traceback (most recent call last):
        3: from /Users/jasonvanderhoof/.rvm/rubies/ruby-2.5.1/bin/irb:11:in `<main>'
        2: from (irb):19
        1: from (irb):13:in `method_missing'
NoMethodError (undefined method `baz' for #<Foo:0x00007f9e74186f40>)
```

orenbm commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,466 @@
+
+The input for this class will be built in each authenticator as it differs
+between them. For simplicity, the class will receive the `constraints` hash
+as an input. We should challenge that to find a better way.
```
@jonahx any idea?

jvanderhoof commented:
@orenbm, this feels like a heavyweight solution to a problem.  What's the one-sentence outcome you're hoping to achieve? Is it something like "simple validation of authenticator attributes"?

izgeri commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,468 @@
+the authentication request was sent from a pod that is in namespace
+`namespace-2` we will fail the request.
+
+In a way, we can define this step as comparing 2 objects of Application
+Identity - one that was extracted from the host and one that was extracted
+from the authentication request. In the example above we will have the
+following Application Identity objects:
+ - Host Application Identity
+   - constraints
+     - namespace: `namespace-1`
+ - Authentication request Application Identity
+   - constraints
+     - namespace: `namespace-2`
```
where does the namespace info come from in authn-k8s when an authentication request is sent from an app with a host identity defined using annotations?

previously the host would have authenticated as `{namespace}/service_account/{service account}`; the namespace / service account would be sent with the request in the form of the host ID (`CONJUR_AUTHN_LOGIN`) and the service_id would be extracted from the request route

now when a host authenticates as `my-host` in the `CONJUR_AUTHN_LOGIN`, where the host was defined in policy using the `{service account}` and `{namespace}` values as host annotations, how does the server extract the "authentication request application identity" values for service account and namespace?

sorry if this is more clearly defined elsewhere, I've just never thought about this aspect of defining hosts by annotations before and your example is raising this question for me

orenbm replied:
As can be seen in [our docs](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/Kubernetes_AppIdentity.htm?Highlight=application%20identity#DefineKubernetesResourcesasDAPhosts) the namespace can be defined in the annotation `authn-k8s/namespace`. 

the host `{namespace}/service_account/{service account}` can be defined as follows:
```
- host 
  id: some-host
  annotations:
    - authn-k8s/namespace: {namespace}
    - authn-k8s/service-account: {service account}
```

The extraction of the values from the request stayed the same, we didn't touch that logic.

izgeri commented:
just wanted to comment that this is a very well thought out and clear document, and it clearly will be valuable to future authenticators to have this centralized logic in place. nice work identifying the issue and drafting this proposal @orenbm!

jonahx commented:
Going to add one long comment with quotes from original instead of inline
GH comments.

## General Comments

This is thorough and attentive to detail.  I appreciate the effort that went
into this, and I'd very much like to see more features developed in this way:
With a thoughtful writeup, and lots of time for iterative feedback from other
devs.

My overall impression is that the design feels too complex what the feature,
essentially, is.  I spent a long time thinking about and examining existing
code to try to figure out why.

## Specifics

I'll start with a few comments about places where the writing of the document
itself wasn't clear enough to me.  This may seem unrelated to the technical
design, but I think they are actually connected.

So first, I want the "Introduction" to frame the general problem more clearly.
Both what it is, and how it relates to Conjur authentication.  Instead, we jump
right in to an assumed knowledge of speficic Conjur authenticators:

> In several authenticators, such as authn-k8s & authn-azure, we require hosts  

We're starting at 1K feet high instead of 10K feet high where we want to be.

> The Application Identity gives the authenticator another layer of
> authentication,

Let's be more precise here.  "Authentication" (generally speaking, not just in
Conjur) is how you verify an identity.  In the most basic case, a client enters
a username/password.  The client's knowledge of the password verifies that he
is the true owner of the username (ie, that's his true identity).

So it's confusing to talk about "identity adding a layer of authentication",
since "authentication" is "the process of verifying identity".

I think what you meant in the above quote is that it is essentially an "alias",
or like a secondary "username" you can use when authenticating.

I think we should explicitly define our terms.  Here are some suggestions:

- "identity" -- the actual thing that you are.  I might call you "Oren" or "Mr.
  Ben Meir" but I always refer to the same identity, the actual flesh and blood
  person who _is_ Oren.
- "identifiers" -- any way of referring to an identity.  "Oren" and "Mr. Ben
  Meir" are identifiers. In our current context, I think it the Conjur host
  would be the identity, and it can be referred to by multiple identifers.

Then you could say stuff like "The annotations are used as another way to
_identify_ a Conjur host" or "The annotations are identifiers of the Conjur
host."

> In the example above, both hosts have the same Application Identity. One
> stores it in its annotation and the other in its ID. However, their Conjur
> identity is not the same.

This was confusing to me.  How can the same identifier refer to two different
hosts?

> However, when validating the Application Identity we currently have no
> generic mechanism, as can be seen in the implementations for the Azure
> Authenticator and Kubernetes Authenticator, where a lot of code is duplicated
> with minor changes.

I looked at these files but wasn't able to immediately see the duplicated
code...  I'd like to know exactly what it is, since this will help me
understand the abstraction you're wanting to create here.

> The Application Identity validation can be roughly split into 3 parts:
> 
> - Extract the Application Identity from the host
> - Validate the Application Identity is defined correctly
> - Validate the Application Identity matches the authentication request

```
class ApplicationIdentity
  ...
end
```

Couple of issues here:

First, this class is basically an `AnnotationRepo` -- that is, it's a just a
wrapper over the database that (typically) returns domain obects (in this case,
hashes).  Here the wrapper is doing 2 things: logging, and building the
annotation path for you (by adding the full prefix with the `service_id`).

But an `AnnotationRepo` is not an abstraction you'd repeat for every 
type of application identity.  You'd just have one of them.

> Validate the Application Identity is defined correctly

`ValidateAzureAnnotations` shoudn't be a `CommandClass`.

Remember, `CommandClass` should be used to implement "use case objects"
(sometimes called "service" objects).  It's where the logic of multiple things
working together lives, so typically you'll only need one when you have 2 or
more dependencies.

Really `ValidateAzureAnnotations` has only pure functional logic.  It can be impleented as a pure value object or pure function: 

You give it some annotations, it returns if they're valid or not.

It can include validation messages, of course, but it won't actually raise any
errors or do any logging.  Just pure input/output.  The error and logging would
be done by the `CommandClass` that calls it.

Here's the big lesson: Pure functions / value objects are the easiest things to
understand and reason about.  Not to mention test (no mocks needed like  there
are when testing a `CommandClass`!).  Therefore we want to push as much of our
logic into them as possible.

The `CommandClass` pieces which wire things together, then, can remain lean and
just delegate to our pure functional code.

> The following lists of annotation names
> - permitted
> - required
> - non-permitted combinations

I wouldn't abstract on this quite yet.  Future authenticators may have
requirements that don't fit into this pattern.  Instead, per above: Make the
API just "take in annotation names, return a validation result".

> Validate the Application Identity matches the authentication request

Before moving on here, let's retrospect on what's happened so far.  Because
conceptually, it's interesting.  Really all we've done is pull a bunch of 
stuff out of our own database and validate it.

Now that's odd... you might wonder why do we have to validate our own data?
Don't you normally validate data _before_ you stick in a DB?

What's happening is here is that we're missing data types.  We're using
free-form annotations as a way to create ad-hoc types that might otherwise be
stored in their own tables.

What are those types?

They are different _kinds_ of hosts.  We don't just have _hosts_.  We now have
_Azure hosts_ and _OIDC hosts_ and so on.  But they are currently  _invisible
types_... they exist only implicitly.  

We very likely want a design that makes these explicit.

Something like this...

A generic `HostRepo` (see my comments above on `AnnotationRepo`), which houses 
the generic annotation logic (just fetching them), and then an `AzureHostRepo` and `OIDCHostRepo`, etc... whose only job is to query the database and return  
fully-formed, validated `AzureHost` and `OIDCHost` objects that work just like 
we want.  Or return an error, if the requesed host cannot be coerced into one of  these types.

Because that's what all this application identity validation is really about...
that's the underlying concept here.  Is the host id you're asking for actuall
an _Azure_ host, or are its annotations inconsistent with that type of host?

Now when we validate the request the implementation detail of looping over
constraints can go away...

What you're really asking is something like: `AzureHost.matches?(request)`

This is just the high-level idea, but think that abstraction would simplify the
code a lot and match how we're naturally thinking about things.

orenbm commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,464 @@
+generic mechanism, as can be seen in the implementations for the [Azure
+Authenticator](/app/domain/authentication/authn_azure/validate_application_identity.rb) and 
+[Kubernetes Authenticator](/app/domain/authentication/authn_k8s/validate_application_identity.rb), where a lot of code is duplicated with
+ minor changes. 
```
the duplicated code is (showing snippets from azure but they appear also in k8s):
```
def initialize(role_annotations:, service_id:, logger:)
        @role_annotations = role_annotations
        @service_id       = service_id
        @logger           = logger
      end

      def constraints
        @constraints ||= {
          subscription_id:          constraint_value("subscription-id"),
          resource_group:           constraint_value("resource-group"),
          user_assigned_identity:   constraint_value("user-assigned-identity"),
          system_assigned_identity: constraint_value("system-assigned-identity")
        }.compact
      end

# check the `service-id` specific constraint first to be more granular
      def constraint_value constraint_name
        annotation_value("authn-azure/#{@service_id}/#{constraint_name}") ||
          annotation_value("authn-azure/#{constraint_name}")
      end

      def annotation_value name
        annotation = @role_annotations.find { |a| a.values[:name] == name }

        # return the value of the annotation if it exists, nil otherwise
        if annotation
          @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(name))
          annotation[:value]
        end
      end

# validating that the annotations listed for the Conjur resource align with the permitted Azure constraints
      def validate_azure_annotations_are_permitted
        validate_prefixed_permitted_annotations("authn-azure/")
        validate_prefixed_permitted_annotations("authn-azure/#{@service_id}/")
      end

      # check if annotations with the given prefix is part of the permitted list
      def validate_prefixed_permitted_annotations prefix
        @logger.debug(LogMessages::Authentication::ValidatingAnnotationsWithPrefix.new(prefix))

        prefixed_annotations(prefix).each do |annotation|
          annotation_name = annotation[:name]
          next if prefixed_permitted_constraints(prefix).include?(annotation_name)
          raise Err::ConstraintNotSupported.new(annotation_name.gsub(prefix, ""), permitted_constraints)
        end
      end

      def prefixed_annotations prefix
        @role_annotations.select do |a|
          annotation_name = a.values[:name]

          annotation_name.start_with?(prefix) &&
            # verify we take only annotations from the same level
            annotation_name.split('/').length == prefix.split('/').length + 1
        end
      end

      # add prefix to all permitted constraints
      def prefixed_permitted_constraints prefix
        permitted_constraints.map { |k| "#{prefix}#{k}" }
      end

      def permitted_constraints
        @permitted_constraints ||= %w(
          subscription-id resource-group user-assigned-identity system-assigned-identity
        )
      end

 def validate_constraint_exists constraint
        unless application_identity.constraints[constraint]
          raise Err::RoleMissingConstraint, annotation_type_constraint(constraint)
        end
      end

 # validates that the application identity doesn't include logical constraint
      # combinations (e.g user_assigned_identity & system_assigned_identity)
      def validate_constraint_combinations
        identifiers = %i(user_assigned_identity system_assigned_identity)

        identifiers_constraints = application_identity.constraints.keys & identifiers
        unless identifiers_constraints.length <= 1
          raise Errors::Authentication::IllegalConstraintCombinations, annotation_type_constraints(identifiers_constraints)
        end
      end

def validate_token_identity_matches_annotations
        application_identity.constraints.each do |constraint|
          annotation_type  = constraint[0].to_s
          annotation_value = constraint[1]
          unless annotation_value == @token_identity[annotation_type.to_sym]
            raise Err::InvalidApplicationIdentity, annotation_type_constraint(annotation_type)
          end
        end
        @logger.debug(Log::ValidatedApplicationIdentity.new)
      end

      def annotation_type_constraints constraints
        constraints.map { |constraint| annotation_type_constraint(constraint) }
      end

      # converts the constraint to be in annotation style (e.g resource-group
      # instead of resource_group) to enhance supportability
      def annotation_type_constraint constraint
        constraint.to_s.tr('_', '-')
      end
```

This is pretty much the whole code of authn-azure's application identity validation and it is implemented also in authn-k8s

orenbm commented:
> @orenbm, this feels like a heavyweight solution to a problem. What's the one-sentence outcome you're hoping to achieve? Is it something like "simple validation of authenticator attributes"?

@jvanderhoof We already have this validation in Conjur, but each authenticator has its on implementation for it although the code is identical in a huge part of it. 

What I'm hoping to achieve is a generic mechanism to validate the application identity so that future authenticators will have minimal implementation, and it can leverage from the general code.

This is the same as we have in authenticators, where there is a general flow (check the security, audit, create an access token) and a specific part that verifies the credentials.

orenbm commented:
thanks @jonahx for the thorough response. I'll comment here regarding your comments:

> So first, I want the "Introduction" to frame the general problem more clearly.

Noted. I assumed some knowledge regarding our authenticators and jumped into introducing the Application Identity. @sigalsax  said that she had an easy time joining my train of thought but she is more familiar with this area.

> I think what you meant in the above quote is that it is essentially an "alias",
or like a secondary "username" you can use when authenticating.

That is actually incorrect. The Application Identity is a layer of authorization (as opposing to a layer of authentication i called it earlier). It limits the host's authentication ability. For example, the host:
```
   - !host
      id: some-host
      annotations:
        authn-k8s/namespace: some-namespace
```
can authenticate to Conjur using the authn-k8s, only from the namespace `some-namespace` and not from another.

The Application Identity is not an alias or a secondary username, but a layer of authorization. I think that it is clearer now how 2 Conjur hosts can have the same Application Identity:
 ```
- !host
  id: test-app-service-account
  annotations:
    authn-k8s/namespace: some-namespace
    authn-k8s/service-account: some-service-account

- !host
  id: some-namespace/service_account/some-service-account
```
One defines it in the ID and one defines it in the annotations but both hosts can authenticate with Conjur from a pod in the namespace `some-namespace` and that has the service-account `some-service-account`.

> I looked at these files but wasn't able to immediately see the duplicated
code

I addressed this question [here](https://github.com/cyberark/conjur/pull/1479#discussion_r426560275). As you can see it is most of the code. The only difference is that the k8s authenticator also handles an Application Identity defined in the ID.

> ValidateAzureAnnotations shoudn't be a CommandClass.

I had this debate with myself when we implemented this in authn-azure. It seemed better to split it out at the time but i can't disagree with you. I'll take this note if we implement the generic validator.

> I wouldn't abstract on this quite yet. Future authenticators may have
requirements that don't fit into this pattern. Instead, per above: Make the
API just "take in annotation names, return a validation result".

I don't really understand your suggestion here. Can you elaborate?

> Now that's odd... you might wonder why do we have to validate our own data?
Don't you normally validate data before you stick it in a DB?

What's happening here is that we're missing data types. We're using
free-form annotations as a way to create ad-hoc types that might otherwise be
stored in their own tables.

You are totally correct and we addressed this issue in the past. Ideally, these validations would occur upon policy load. However, we didn't get the time to redesign this and add a new data-type instead of using the current mechanism of annotations.

> They are different kinds of hosts. We don't just have hosts. We now have
Azure hosts and OIDC hosts and so on. But they are currently invisible
types... they exist only implicitly.

While this is conceptually correct, we also have a requirement to let hosts authenticate with multiple authenticators.

eranha commented on design/authenticators/application-identity/application-identity.md:
```diff
@@ -0,0 +1,468 @@
+
+### Defining Application Identity Validation
+
+The Application Identity validation can be roughly split into 3 parts:
```
It'a actually three responsibilities.

jonahx commented:
I re-read the k8's `application_identity.rb` to get more context for my answer,
and I noticed a few things...

You had asked me a while back what you could do to improve:

```ruby 
def constraints
  @constraints ||= {
    pod:               constraint_value("pod"),
    service_account:   constraint_value("service_account"),
    deployment:        constraint_value("deployment"),
    deployment_config: constraint_value("deployment_config"),
    stateful_set:      constraint_value("stateful_set")
  }.compact
end
```

You'd complained that, since the attributes being returned could be different
depending on the host, you had to use a hash rather than a fixed object.  That
is, you recognized that it wasn't returning a _fixed type_ and that (correctly)
didn't sit right with you.

The answer to your question is found in how we're _using_ the above method.  I
did a "Find Usages" in RubyMine, and there were only 2 relevant non-spec
results.  Here it is being used in `ValidateApplicationIdentity`:

```ruby
def validate_pod_metadata
  application_identity.constraints.each do |constraint|
    resource_type   = constraint[0].to_s
    resource_name   = constraint[1]
    resource_object = k8s_resource_object(
      resource_type,
      resource_name,
      application_identity.namespace
    )
```

That is, conceptually, `constraints` really returns an a list of K8s resources.
That's your missing object.  If you add `K8sResource` value object,
`constraints` can have a fixed return type: `[<K8sResource>]` and the method
`constraints` can be renamed to reflect that.

Note: You'll need to create a new value type for `K8sResource` since the
existing `k8s_resource_object` is something else: It interacts with the K8s API
and returns results from that.

Additionally, I think you can probably rewrite that `ApplicationIdentity` class
and remove 1/3 or more of the code.

I would suggest doing this before contining the "Application Identity"
abstraction being proposed here.  As I mentioned above, "Application Identity"
is not really what this is about.

My suggestion is to return to first principles:

1. Write down in a plain English bullet list, at a high level, what validations
we're doing in `ApplicationIdentity`.
2. Make sure that code -- both the names, the number methods, and their
structure -- reflect what you wrote in step 1.

Remember the _goal_ is not to have many one-line methods.  The goal is to make
your high-level logic clear and suppress extraneous detail.  One line private
methods can often help with this. But if... 

1. They aren't named carefully or 
2. If they're not broken down at the right level of granularity or
3. If keeping the call chain in your head becomes harder than following a single, longer method body 

Then it's time to rethink.

Getting back to the main project now... My advice here is more speculative 
since I'm not 100% sure I understand all the requirements...

Given my current understanding, I think what all this is really about is
validating hosts/applications.  Different hosts have different  sets of
properties, which are stored in annotations (or paths) and we need to validate
them after we pull them out of the database.  The basic pattern (correct
me if I'm wrong here) is something like:

1. Query the database using a host_id (or other application identifer?)
2. Get back a bunch of annotations
3. Validate those annotations are valid since we don't do so when inserting
   them which we should but that's a separate discussion.
4. Turn that set of annotations into a semantically meaningful object like a
   k8s resource (this part we're not doing yet)
5. Do something with those objects (that's not part of the scope of this document)


```ruby
# Alternative name: ApplicationRepo?
class HostRepo
  def k8_application(host_id)
    # 1. Does db query for the annotations
    # 2. Delegates to the validation logic currently in 
    #    ApplicationIdentity
    # 3. Returns [K8sResource] per above
  end

  # Same for azure and others
end
```

Then `ApplictionIdentity` get renamed to something like `K8Application` or
`K8Host` or `K8sResources` (I'm a little hazy on the correct terminology here,
but I'm guessing a K8sApplication is made up of a set of K8s resources?).  That
is, we create our K8sApplication  from the annotations, and they get validated
as part of that creation:

```ruby
class K8sApplication
  def initialize(host_id:, host_annotations:, service_id:)
    # etc...
  end
end
```

orenbm commented:
thanks @jonahx ! please review [this PR](https://github.com/cyberark/conjur/pull/1580). I think that after we merge it, it will be much easier to generalize the validations.

orenbm commented:
this effort is at a halt at the moment and may be continues in the future. Until then, i am closing this PR
